pymc_experimental/model_builder.py

```diff
@@ -0,0 +1,129 @@
+import pymc as pm
+import arviz
+import pandas as pd
+from pathlib import Path
+import numpy as np
+from typing import Dict
+import cloudpickle
+import arviz as az
+import hashlib
+
+
+class ModelBuilder(pm.Model):
+
+    _model_type = "BaseClass"
+    version = "None"
+
+    def __init__(self, data: pd.DataFrame, model_config: Dict, sampler_config: Dict):
+        super().__init__()
+        self.data = data  # input and output data
+        self.model_config = model_config  # parameters for priors etc.
+        self.sample_config = sampler_config  # parameters for sampling
+
+        self.idata = None
+
+    def _build(self):
+        raise NotImplementedError
+
+    def _data_setter(self, data: pd.DataFrame, x_only: bool = True):
+
+        raise NotImplementedError
+
+    @classmethod
+    def create_sample_input(cls):
+        """
+        Needs to be implemented by the user in the inherited class.
+        Returns examples for data, model_config, samples_config.
+        This is useful for understanding the required 
+        data structures for the user model.
+        """
+        raise NotImplementedError
+
+    def build(self):
+        with self:
+            self._build()
+
+    def save(self, file_prefix, filepath, save_model=True, save_idata=True):
+        if save_idata:
+            file = Path(filepath + str(file_prefix) + ".nc")
+            self.idata.to_netcdf(file)
+        if save_model:
+            filepath = Path(str(filepath) + str(file_prefix) + ".pickle")
+            Model = cloudpickle.dumps(self)
+            file = open(filepath, "wb")
+            file.write(Model)
+        self.saved = True
+
+    @classmethod
+    def load(cls, file_prefix, filepath, load_model=True, laod_idata=True):
+        file = Path(str(filepath) + str(file_prefix) + ".pickle")
+        file = open(file, "rb")
+        self = cloudpickle.loads(file.read())
+        filepath = Path(str(filepath) + str(file_prefix) + ".nc")
+        data = az.from_netcdf(filepath)
+        self.idata = data
+        return self
+
+    # fit and predict methods
+    def fit(self, data: pd.DataFrame = None):
+        if data is not None:
+            self.data = data
```

Comment:
What if the model is built for data that is not a single DataFrame?
At least for my models this is almost always the case.

How about making `data` a dictionary that can take DataFrames or NumPy arrays?  (`Dict[Str, Union[np.ndarray, pd.DataFrame, pd.Series]]`